src/TypeInference/GenerateConstraints/Kinds.hs

```diff
@@ -51,6 +55,259 @@ newKVar = do
+annotTy (RST.TyData loc pol xtors) = do 
+  let xtnms = map RST.sig_name xtors
+  xtorKinds <- mapM lookupXtorKind xtnms
+  let allEq = compXtorKinds xtorKinds  
+  case allEq of 
+    Nothing -> throwOtherError loc ["Not all xtors have the same return kind"]
+    Just mk -> do 
+      xtors' <- mapM (\x -> lookupXtorSig loc x pol) xtnms
+      return $ TST.TyData loc pol mk xtors' 
+  where 
+    compXtorKinds :: [MonoKind] -> Maybe MonoKind
+    compXtorKinds [] = Nothing 
+    compXtorKinds [mk] = Just mk
+    compXtorKinds (xtor1:xtor2:rst) = if xtor1==xtor2 then compXtorKinds (xtor2:rst) else Nothing
+annotTy (RST.TyCodata loc pol xtors) = do 
+  let xtnms = map RST.sig_name xtors
+  xtorKinds <- mapM lookupXtorKind xtnms
+  let allEq = compXtorKinds xtorKinds  
+  case allEq of 
+    Nothing -> throwOtherError loc ["Not all xtors have the same return kind"]
+    Just mk -> do 
+      xtors' <- mapM (\x -> lookupXtorSig loc x (RST.flipPolarityRep pol)) xtnms
+      return $ TST.TyCodata loc pol mk xtors' 
+  where 
+    compXtorKinds :: [MonoKind] -> Maybe MonoKind
+    compXtorKinds [] = Nothing 
+    compXtorKinds [mk] = Just mk
+    compXtorKinds (xtor1:xtor2:rst) = if xtor1==xtor2 then compXtorKinds (xtor2:rst) else Nothing
+annotTy (RST.TyDataRefined loc pol tyn xtors) =  do 
+  xtors' <- mapM annotXtor xtors
+  tyn' <- gets declTyName
+  if tyn == tyn' then do
+    polyknd <- gets declKind
+    return $ TST.TyDataRefined loc pol (CBox $ returnKind polyknd) tyn xtors' 
+  else do 
+    decl <- lookupTypeName loc tyn
+    return $ TST.TyDataRefined loc pol (CBox $ returnKind $ TST.data_kind decl) tyn xtors' 
+annotTy (RST.TyCodataRefined loc pol tyn xtors) = do 
+  xtors' <- mapM annotXtor xtors
+  tyn' <- gets declTyName
+  if tyn == tyn' then do 
+    polyknd <- gets declKind
+    return $ TST.TyCodataRefined loc pol (CBox $ returnKind polyknd) tyn xtors'
+  else do
+    decl <- lookupTypeName loc tyn
+    return $ TST.TyCodataRefined loc pol (CBox $ returnKind (TST.data_kind decl)) tyn xtors'
+annotTy (RST.TyNominal loc pol tyn vartys) = do 
+  tyn' <- gets declTyName
+  vartys' <- mapM annotVarTy vartys
+  if tyn == tyn' then do
+    polyknd <- gets declKind
+    return $ TST.TyNominal loc pol (CBox $ returnKind polyknd) tyn vartys' 
+  else do
+    decl <- lookupTypeName loc tyn
+    return $ TST.TyNominal loc pol (CBox $ returnKind (TST.data_kind decl)) tyn vartys' 
+annotTy (RST.TySyn loc pol tyn ty) =  do 
+  ty' <- annotTy ty
+  return $ TST.TySyn loc pol tyn ty'
+annotTy (RST.TyBot loc) = do TST.TyBot loc . KindVar <$> newDeclKVar
+annotTy (RST.TyTop loc) = do TST.TyTop loc . KindVar <$> newDeclKVar
+annotTy (RST.TyUnion loc ty1 ty2) = do 
+  ty1' <- annotTy ty1 
+  ty2' <- annotTy ty2
+  let knd = TST.getKind ty1' 
+  if knd == TST.getKind ty2' then 
+    return $ TST.TyUnion loc knd ty1' ty2'
+  else 
+    throwOtherError loc ["Kinds of " <> T.pack ( show ty1' ) <> " and " <> T.pack ( show ty2' ) <> " in union do not match"]
```

Comment:
Das ist glaube ich etwas zu restriktiv, was ist mit `⊤⊓⊤`?
Da würde es einen Fehler geben, obwohl das ok zu sein scheint.
Reicht es hier nicht eventuell, einfach eine neue Variable zu vergeben?
Oder gibt das Probleme, weil für die Declarations keine Constraints erzeugt werden?

Reply:
Genau, ich kann in annotTy keine variablen erstellen, sonst fehlen die später beim constraint solving. Aber wir haben ja auch gesagt Top und Bottom dürfen in den Declarations gar nicht stehen, also glaube ich das kann nie vorkommen